### ISSUE TITLE
ES.30 incomplete Note

### ISSUE THREAD

**Issue by omtcyfz** [isocpp/CppCoreGuidelines#748]
[ES.30: Don’t use macros for program text manipulation](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#a-nameres-macrosaes30-dont-use-macros-for-program-text-manipulation) states that macros should be banned from everything except of `#ifdefs` and configuration control.

However, I disagree with that. Macros are useful for the purpose of assertions and logging. I think that "configuration control" doesn't cover that.

The original Note
> This rule does not ban the use of macros for “configuration control” use in #ifdefs, etc.

**Comment by rianquinn** [comment]
I agree with this. Unlike Swift, C++ doesn't have a way of using \_\_LINE\_\_ as a default argument to a function, so often times logging functions that would like to show the line #, and function name have to be implemented with macros, and it's only done because this feature is missing in C++. 

For example:

The following is what would be needed to remove the need for macros (what Swift does):
```
void log_error(const char *msg, const char *func = __PRETTY_FUNCTION__, int line = __LINE__);
```

What needs to be done instead:
```
#define log_error(a) log_error_with_args(a, __PRETTY_FUNCTION__, __LINE__)
void log_error_with_args(const char *msg, const char *func, int line);
```

In both cases the following is how it would be used:
```
log_error("the answer was not 42");
```

IMO, there is nothing wrong with the above code, and is a pattern that helps more than hurts, but once again, is only needed because C++ has no way of supporting this functionality without macros.

**Comment by MikeGitb** [comment]
I actually see three valid use cases for macros:
- conditional compilation: #ifdef ...
- logging 
- generic programming: Sometimes you want to repeat some fairly complicated definition / expression and e.g. only change the **name** of a function or type. Or you want to have a string that exactly matches an Identifier. In general, they sometimes allow you to more strictly follow the DRY principle without resorting to external tools and scripts that would generate that code for you.

In the last case those macros can (and imho should) be undefined right after usage.

**Comment by AndrewPardoe** [comment]
Bjarne notes that we should reference the intrinsics for __LINE__ and __FILE__ that are proposed for C++17. We really don't want to encourage macros because it's always a camel's nose. This guideline is aimed at future code.

**Comment by jwakely** [comment]
The type [`std::experimental::source_location`](https://rawgit.com/cplusplus/fundamentals-ts/v2/fundamentals-ts.html#reflection.src_loc) makes most uses of `__FILE__` and `__LINE__` redundant for logging purposes.
